File: haskell/pangram/Iteration2/src/Pangram.hs
module Pangram (isPangram) where

import Data.Char
import qualified Data.Set as Set

isPangram :: String -> Bool
isPangram text = sortedChars == checksum
    where
        allowedChars = filter isAllowedChar text
        distinctChars   = Set.fromList allowedChars
        lowercaseChars  = Set.map toLower distinctChars
        sortedChars     = Set.toAscList lowercaseChars

checksum :: String
checksum = ['a'..'z']

isAllowedChar :: Char -> Bool
isAllowedChar c = isAlpha c && isAscii c

File: c/pangram/Iteration2/pangram.c
#include <stddef.h>
#include <ctype.h>

#include "pangram.h"

const unsigned long all_letters_mask = (1UL << 26) - 1;

bool is_pangram(const char *sentence) {
    if (sentence == NULL) return false;
    if (*sentence == '\0') return false;

    unsigned long letterFlags = 0;
    
    for (const char* p = sentence; *p != '\0'; p++) {
        char c = *p;
        int letterIndex = 0;
        
        if (isalpha(c))
        {
            letterIndex = tolower(c) - 'a';
            letterFlags |= (1UL << letterIndex);
        }
    }
    
    return (letterFlags & all_letters_mask) == all_letters_mask;
}
